diag/diagnostics.go

```diff
@@ -0,0 +1,97 @@
+package diag
+
+import (
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// Diagnostics represents a collection of diagnostics.
+//
+// While this collection is ordered, the order is not guaranteed as reliable
+// or consistent.
+type Diagnostics []Diagnostic
+
+// AddAttributeError adds a generic attribute error diagnostic to the collection.
+func (diags *Diagnostics) AddAttributeError(path *tftypes.AttributePath, summary string, detail string) {
+	diags.Append(NewAttributeErrorDiagnostic(path, summary, detail))
+}
+
+// AddAttributeWarning adds a generic attribute warning diagnostic to the collection.
+func (diags *Diagnostics) AddAttributeWarning(path *tftypes.AttributePath, summary string, detail string) {
+	diags.Append(NewAttributeWarningDiagnostic(path, summary, detail))
+}
+
+// AddError adds a generic error diagnostic to the collection.
+func (diags *Diagnostics) AddError(summary string, detail string) {
+	diags.Append(NewErrorDiagnostic(summary, detail))
+}
+
+// AddWarning adds a generic warning diagnostic to the collection.
+func (diags *Diagnostics) AddWarning(summary string, detail string) {
+	diags.Append(NewWarningDiagnostic(summary, detail))
+}
+
+// Append adds non-empty and non-duplicate diagnostics to the collection.
+func (diags *Diagnostics) Append(in ...Diagnostic) {
```

Comment:
Does this work with `Diagnostics` being returned? E.g.

```go
diags := doTheThing() // diags is now of type Diagnostics
resp.Diagnostics.Append(diags...) // does this work? diags is of type Diagnostics, not []Diagnostic, but Go may convert for us?
```

Reply:
`Diagnostics` is just a type alias for `[]Diagnostic` so `diags...` is equivalent to `...Diagnostic`. See `tfsdk/serve.go` updates for this in action (compiles and unit tests fine).